### PR TITLE
Triangular solver local restructuring

### PR DESCRIPTION
**Pull request by Sely85**
Local implementation of triangular solver modified to adapt to new API and structure directory. 

New structure is:

```
include/dlaf
    solver.h
    solver/
        mc.h
        triangular/
            mc.tpp
            mc/
                triangular_LLN.h
                triangular_LLT.h
                triangular_LUN.h
                triangular_LUT.h
                triangular_RLN.h
                triangular_RLT.h
                triangular_RUN.h
                triangular_RUT.h  
```

### CODE REVIEW

File: test/unit/solver/mc/test_triangular.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#include "dlaf/solver/mc.h"
+
+#include <exception>
+#include <functional>
+#include <sstream>
+#include <tuple>
+#include "../../test_blas_tile/test_trsm.h"
```
[the remark this answers is not in the document]
Comment (Sely85) [reply test/unit/solver/mc/test_triangular.cpp]: I've opened issue #151

File: include/dlaf/solver/triangular/mc/triangular_LLN.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include <hpx/include/parallel_executors.hpp>
+#include <hpx/include/threads.hpp>
+
+#include "dlaf/blas_tile.h"
+#include "dlaf/common/index2d.h"
+#include "dlaf/common/pipeline.h"
```
[the remark this answers is not in the document]
Comment (Sely85) [reply include/dlaf/solver/triangular/mc/triangular_LLN.h]: Right, sorry

File: include/dlaf/solver/triangular/mc/triangular_LLN.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include <hpx/include/parallel_executors.hpp>
+#include <hpx/include/threads.hpp>
```
[the remark this answers is not in the document]
Comment (Sely85) [reply include/dlaf/solver/triangular/mc/triangular_LLN.h]: Substituting these two lines with the suggested include results in the following error: 
`error: ‘pool_executor’ is not a member of ‘hpx::threads::executors’`.

File: test/unit/CMakeLists.txt [hunk 4]
```diff
@@ -21,4 +21,4 @@ add_subdirectory(communication)
 add_subdirectory(factorization)
 add_subdirectory(matrix)
 add_subdirectory(memory)
-add_subdirectory(mc)
+add_subdirectory(solver)
```
[the remark this answers is not in the document]
Comment (rasolca) [reply test/unit/CMakeLists.txt]: I agree. What about adding a small comment before each group to make the separation clear?
```
# ??? (any suggestion?)
add_subdirectory(matrix)
add_subdirectory(memory)

# Operations
add_subdirectory(factorization)
add_subdirectory(solver)
```

File: include/dlaf/solver/triangular/mc/triangular_LLN.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/solver/triangular/mc/triangular_LLN.h]: Unfortunately `hpx/hpx.hpp` doesn't include everything... AFAIR both these headers are required.

File: test/unit/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (albestro) [reply test/unit/CMakeLists.txt]: I completely agree. For now, I would suggest to have just two categories, one for the **generic** library (e.g. memory, matrix, common, ...), and one for the **operations**.

[the remark this answers is not in the document]
Comment (Sely85) [reply test/unit/CMakeLists.txt]: Looks good.

File: include/dlaf/solver/triangular/mc/triangular_LLN.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      hpx::threads::executors::pool_executor("default", hpx::threads::thread_priority_default);
+
+  SizeType m = mat_b.nrTiles().rows();
+  SizeType n = mat_b.nrTiles().cols();
+
+  for (SizeType k = 0; k < m; ++k) {
+    for (SizeType j = 0; j < n; ++j) {
+      auto kj = LocalTileIndex{k, j};
+
+      // Triangular solve of the first tile
```
Comment (rasolca) [inline include/dlaf/solver/triangular/mc/triangular_LLN.h]:
```suggestion
      // Triangular solve of k-th row Panel of B
```

File: include/dlaf/solver/triangular/mc/triangular_LLN.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      hpx::dataflow(executor_hp, hpx::util::unwrapping(tile::trsm<T, Device::CPU>), Left, Lower, NoTrans,
+                    diag, alpha, mat_a.read(LocalTileIndex{k, k}), std::move(mat_b(kj)));
+
+      for (SizeType i = k + 1; i < m; ++i) {
+        // Choose queue priority
+        auto trailing_executor = (i == k + 1) ? executor_hp : executor_normal;
+        auto beta = static_cast<T>(-1.0) / alpha;
+        // Matrix multiplication
```
Comment (rasolca) [inline include/dlaf/solver/triangular/mc/triangular_LLN.h]:
```suggestion
        // Update trailing matrix of B
```

File: include/dlaf/solver/triangular/mc/triangular_LLT.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      hpx::threads::executors::pool_executor("default", hpx::threads::thread_priority_high);
+
+  // Set up executor on the default queue with default priority.
+  hpx::threads::scheduled_executor executor_normal =
+      hpx::threads::executors::pool_executor("default", hpx::threads::thread_priority_default);
+
+  SizeType m = mat_b.nrTiles().rows();
+  SizeType n = mat_b.nrTiles().cols();
+
+  // Loop on rows
```
Comment (rasolca) [inline include/dlaf/solver/triangular/mc/triangular_LLT.h]: this comment should be removed

File: include/dlaf/solver/triangular/mc/triangular_LLT.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  for (SizeType k = m - 1; k > -1; --k) {
+    // Loop on cols
```
Comment (rasolca) [inline include/dlaf/solver/triangular/mc/triangular_LLT.h]: this comment should be removed

File: include/dlaf/solver/triangular/mc/triangular_RLN.h [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  SizeType m = mat_b.nrTiles().rows();
+  SizeType n = mat_b.nrTiles().cols();
+
+  // Loop on cols
+  for (SizeType k = n - 1; k > -1; --k) {
+    // Loop on rows
+    for (SizeType i = m - 1; i > -1; --i) {
+      auto ik = LocalTileIndex{i, k};
+
+      // Triangular solve of k-th row Panel of B
```
Comment (rasolca) [inline include/dlaf/solver/triangular/mc/triangular_RLN.h]: When side is Right, the panel is a col panel.

File: include/dlaf/solver/triangular/mc/triangular_LLT.h (hunk 7 above)
Comment (rasolca) [reply include/dlaf/solver/triangular/mc/triangular_LLT.h]: Note: check the other cases as well.

**Pull request merged**